### ISSUE TITLE
Problem compiling libwebsockets with wolfssl-3.9.10

### ISSUE THREAD

**Issue by lyntongrice** [warmcat/libwebsockets#741]
Hi there,

Forgive me if I have something stupid here but I have compiled wolfssl (wolfssl-3.9.10) with the following on my Linux OpenSuse box.

./configure --prefix=/opt/wolfssl/ CFLAGS="-m64 -fPIC" --enable-sniffer --enable-static --enable-opensslextra
make 
make install

That all works file, now I want to "cmake" libwebsockets (downloaded last week) with the following:

cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/opt/libwebsockets/ -DLIB_SUFFIX=64 -DLWS_USE_CYASSL=0 -DLWS_USE_WOLFSSL=1 -DLWS_WOLFSSL_INCLUDE_DIRS=/opt/wolfssl/include -DLWS_WOLFSSL_LIBRARIES=/opt/wolfssl/lib64/libwolfssl.so

 And then run "make" I keep getting the following issue:

```
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 11%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 12%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 16%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 18%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c: In function ‘lws_ssl_server_name_cb’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:160:10: error: ‘SSL_TLSEXT_ERR_NOACK’ undeclared (first use in this function)
   return SSL_TLSEXT_ERR_NOACK;
          ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:160:10: note: each undeclared identifier is reported only once for each function it appears in
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:163:6: error: implicit declaration of function ‘SSL_get_SSL_CTX’ [-Werror=implicit-function-declaration]
      SSL_get_SSL_CTX(ssl),
      ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:164:6: error: passing argument 1 of ‘wolfSSL_CTX_get_ex_data’ makes pointer from integer without a cast [-Werror]
      openssl_SSL_CTX_private_data_index);
      ^
In file included from /opt/kamoebagateway/wolfssl/include/wolfssl/openssl/ssl.h:33:0,
                 from /home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/private-libwebsockets.h:263,
                 from /home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:22:
/opt/kamoebagateway/wolfssl/include/wolfssl/ssl.h:884:19: note: expected ‘const struct WOLFSSL_CTX *’ but argument is of type ‘int’
 WOLFSSL_API void* wolfSSL_CTX_get_ex_data(const WOLFSSL_CTX*, int);
                   ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:173:19: error: comparison between pointer and integer [-Werror]
   if (vh->ssl_ctx == SSL_get_SSL_CTX(ssl))
                   ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:181:2: error: implicit declaration of function ‘SSL_get_servername’ [-Werror=implicit-function-declaration]
  servername = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
  ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:181:39: error: ‘TLSEXT_NAMETYPE_host_name’ undeclared (first use in this function)
  servername = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
                                       ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:188:4: error: implicit declaration of function ‘SSL_set_SSL_CTX’ [-Werror=implicit-function-declaration]
    SSL_set_SSL_CTX(ssl, vhost->ssl_ctx);
    ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:189:11: error: ‘SSL_TLSEXT_ERR_OK’ undeclared (first use in this function)
    return SSL_TLSEXT_ERR_OK;
           ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c: In function ‘lws_context_init_server_ssl’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:364:2: error: implicit declaration of function ‘SSL_CTX_set_tlsext_servername_callback’ [-Werror=implicit-function-declaration]
  SSL_CTX_set_tlsext_servername_callback(vhost->ssl_ctx,
  ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:395:3: error: implicit declaration of function ‘SSL_CTX_clear_options’ [-Werror=implicit-function-declaration]
   SSL_CTX_clear_options(vhost->ssl_ctx, info->ssl_options_clear);
   ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c: In function ‘lws_ssl_server_name_cb’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:195:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1: all warnings being treated as errors
CMakeFiles/websockets.dir/build.make:398: recipe for target 'CMakeFiles/websockets.dir/lib/ssl-server.c.o' failed
make[2]: *** [CMakeFiles/websockets.dir/lib/ssl-server.c.o] Error 1
CMakeFiles/Makefile2:395: recipe for target 'CMakeFiles/websockets.dir/all' failed
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

Seems it cannot find "ssl.h" for some reason I think?

I tried to edit the CMakeList.txt to force in the following:

```
foreach(inc ${WOLFSSL_INCLUDE_DIRS})
                                include_directories("${inc}" "${inc}/wolfssl" "${inc}/wolfssl/openssl" "${inc}/wolfssl/wolfcrypt")
                                message("Include wolfssl header files: ${inc} ${inc}/wolfssl ${inc}/wolfssl/openssl ${inc}/wolfssl/wolfcrypt")
                        endforeach()
```

But that also did not work.

Any idea how I can fix this easily? Or change something to "cmake" command?

Thanks for the help

Lynton

**Comment by lws-team** [comment]
Hm it has no problems finding ssl.h.  Otherwise it dies more spectacularly.

This looks very similar to the "ancient OpenSSL" tls_ext problem that's also an outstanding issue.  I provided that guy a patch thismorning.

It's true that SSL_TLSEXT_ERR_NOACK doesn't exist anywhere in the WolfSSL headers?

**Comment by lyntongrice** [comment]
Hi there,

I can only see SSL_TLSEXT_ERR_NOACK in /opt/wolfssl/include/wolfssl/openssl/ssl.h like follows:

#define SSL_TLSEXT_ERR_NOACK                 alert_warning

Any idea what I can do here?

Thanks for the help

Lynton

**Comment by lws-team** [comment]
These missing bits are about SNI on an SSL-enabled server... if that what you are trying to do?  Because it seems the necessary bits are not there in WolfSSL.

If you don't care about that, "fixing it" or some amount of it anyway is likely to just need this patch

```
diff --git a/lib/private-libwebsockets.h b/lib/private-libwebsockets.h
index cd36c35..325b289 100644
--- a/lib/private-libwebsockets.h
+++ b/lib/private-libwebsockets.h
@@ -254,6 +254,7 @@ lws_plat_get_peer_simple(struct lws *wsi, char *name, int namelen);
 #else
 #include <wolfssl/openssl/ssl.h>
 #include <wolfssl/error-ssl.h>
+#define OPENSSL_NO_TLSEXT
 #endif /* not USE_OLD_CYASSL */
 #else
 #if defined(LWS_USE_POLARSSL)
```

**Comment by lyntongrice** [comment]
Hi there.

Sorry for the late response, just tried it adding that one line above like you said and it gets a different error now:

lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 11%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 12%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 16%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 18%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c: In function ‘lws_context_init_server_ssl’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:395:3: **error: implicit declaration of function ‘SSL_CTX_clear_options’ [-Werror=implicit-function-declaration]
   SSL_CTX_clear_options(vhost->ssl_ctx, info->ssl_options_clear);
   ^**
cc1: all warnings being treated as errors
CMakeFiles/websockets.dir/build.make:398: recipe for target 'CMakeFiles/websockets.dir/lib/ssl-server.c.o' failed
make[2]: *** [CMakeFiles/websockets.dir/lib/ssl-server.c.o] Error 1
CMakeFiles/Makefile2:395: recipe for target 'CMakeFiles/websockets.dir/all' failed
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2


I will try work it out myself but if you have any ideas please let me know.

Thanks again for the help
Lynton

**Comment by lyntongrice** [comment]
Hi there,

I may be way off but it seems to be because of the line:

#ifdef HAVE_STUNNEL

in my /opt/wolfssl/include/wolfssl/openssl/ssl.h file.

As the following is only defined in that IF statement:

#define SSL_CTX_clear_options wolfSSL_CTX_clear_options

Let me know what you think?

Thanks

Lynton

**Comment by lyntongrice** [comment]
I did recompile wolfssl with "stunnel" support but still no luck ;-(

./configure --prefix=/opt/jigerbox/wolfssl/ CFLAGS="-m64 -fPIC" --enable-sniffer --enable-static **--enable-stunnel**
make 
make install

**Comment by lyntongrice** [comment]
excuse the small error above, was actually "/opt/wolfssl"

So the current situation is I have compiled wolfssl with STUNNEL support with 👍 
./configure --prefix=/opt/wolfssl/ CFLAGS="-m64 -fPIC" --enable-sniffer --enable-static --enable-stunnel
make
make install

And have run the following:

cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/opt/libwebsockets/ -DLIB_SUFFIX=64 -DLWS_USE_CYASSL=0 -DLWS_USE_WOLFSSL=1 -DLWS_WOLFSSL_INCLUDE_DIRS=/opt/wolfssl/include -DLWS_WOLFSSL_LIBRARIES=/opt/wolfssl/lib/libwolfssl.so

and "make" still give me:

Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 11%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 12%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 16%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 18%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c: In function ‘lws_context_init_server_ssl’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-server.c:395:3: **error: implicit declaration of function ‘SSL_CTX_clear_options’ [-Werror=implicit-function-declaration]
   SSL_CTX_clear_options(vhost->ssl_ctx, info->ssl_options_clear);**
   ^
cc1: all warnings being treated as errors
CMakeFiles/websockets.dir/build.make:398: recipe for target 'CMakeFiles/websockets.dir/lib/ssl-server.c.o' failed
make[2]: *** [CMakeFiles/websockets.dir/lib/ssl-server.c.o] Error 1
CMakeFiles/Makefile2:395: recipe for target 'CMakeFiles/websockets.dir/all' failed
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2

Any ideas?

**Comment by lyntongrice** [comment]
Hi there,

Only way I can get this to compile for now is to "temporarily" comment out the BOLD line below in ssl-server.c (line 395).

if (info->ssl_options_set)
                SSL_CTX_set_options(vhost->ssl_ctx, info->ssl_options_set);
        if (info->ssl_options_clear)
                /***SSL_CTX_clear_options(vhost->ssl_ctx, info->ssl_options_clear);** */

Anything else anyone can think of here?

Thanks 

Lynton

**Comment by lws-team** [comment]
Seems WolfSSL version you have doesn't provide the API.

By coincidence another live "issue" https://github.com/warmcat/libwebsockets/issues/744 cares about the same code, due to ancient OpenSSL version.

It's hard enough trying to track all the permutations of OpenSSL without different libs as well... if the author of that sends a patch we can adapt it to also be operable if WolfSSL selected.

**Comment by lws-team** [comment]
I pushed on master and v2.1-stable a modification to the fix for #744 that should also fix this.

**Comment by lyntongrice** [comment]
Thanks, let me give it a try -- I will report back just now ....thanks

**Comment by lyntongrice** [comment]
Hi there,

Just got the latest from "git clone https://github.com/warmcat/libwebsockets.git" and now get another error, probably something very small?

Let me know what you think or if I should contact wolfssl for them to make some change?

lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 11%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 12%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 16%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 18%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
[ 20%] Building C object CMakeFiles/websockets.dir/lib/ssl-client.c.o
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/**ssl-client.c: In function ‘lws_ssl_client_bio_create’:
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-client.c:142:44: error: ‘OpenSSL_client_verify_callback’ undeclared (first use in this function)
  SSL_set_verify(wsi->ssl, SSL_VERIFY_PEER, OpenSSL_client_verify_callback);**
                                            ^
/home/Lynton/Desktop/KamoebaGateway/libwebsockets/lib/ssl-client.c:142:44: note: each undeclared identifier is reported only once for each function it appears in
CMakeFiles/websockets.dir/build.make:422: recipe for target 'CMakeFiles/websockets.dir/lib/ssl-client.c.o' failed
make[2]: *** [CMakeFiles/websockets.dir/lib/ssl-client.c.o] Error 1
CMakeFiles/Makefile2:395: recipe for target 'CMakeFiles/websockets.dir/all' failed
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build #

**Comment by namowen** [comment]
Hi, try this patch:

```
diff --git a/lib/ssl-client.c b/lib/ssl-client.c
index 0ce654b..8b3cba2 100644
--- a/lib/ssl-client.c
+++ b/lib/ssl-client.c
@@ -138,8 +138,13 @@
 	}
 
 #endif
+
+#ifndef USE_WOLFSSL
+#ifndef USE_OLD_CYASSL
 	/* OpenSSL_client_verify_callback will be called @ SSL_connect() */
 	SSL_set_verify(wsi->ssl, SSL_VERIFY_PEER, OpenSSL_client_verify_callback);
+#endif
+#endif
 
 #ifndef USE_WOLFSSL
 	SSL_set_mode(wsi->ssl,  SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER);

```

**Comment by lyntongrice** [comment]
Hi there,

That seemed to help but now I am getting another issue ;-( Any ideas?

```
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/ssl-client.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/sha-1.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/lws-plat-unix.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/server.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/server-handshake.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/extension.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/extension-permessage-deflate.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/ranges.c.o
[ 11%] Linking C static library lib/libwebsockets.a
[ 30%] Built target websockets
Scanning dependencies of target test-client
[ 31%] Building C object CMakeFiles/test-client.dir/test-server/test-client.c.o
make[2]: *** No rule to make target '/opt/kamoebagateway/wolfssl/lib/libwolfssl.so', needed by 'bin/libwebsockets-test-client'.  Stop.
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/test-client.dir/all' failed
make[1]: *** [CMakeFiles/test-client.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by lws-team** [comment]
In the original cmake you mentioned, you give the path `/opt/wolfssl/lib64/libwolfssl.so`... but in this error it is `/opt/kamoebagateway/wolfssl/lib/libwolfssl.so`... which one actually has the .so file?

**Comment by lyntongrice** [comment]
Hi there,

Sorry for the confusion, the issue still remains as I cleaned everything out now and you can see all my commands below.

BTW: I am using wolfssl-3.9.10

Install wolfssl first

```
./configure --prefix=/opt/kamoebagateway/wolfssl/ CFLAGS="-m64 -fPIC" --enable-sniffer --enable-static 
make 
make install
```

Now try compile libwebsockets from build directory:
`
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/opt/kamoebagateway/libwebsockets/ -DLIB_SUFFIX=64 -DLWS_USE_CYASSL=0 -DLWS_USE_WOLFSSL=1 -DLWS_WOLFSSL_INCLUDE_DIRS=/opt/kamoebagateway/wolfssl/include -DLWS_WOLFSSL_LIBRARIES=/opt/kamoebagateway/wolfssl/lib/libwolfssl.so`

And when I try make I get:
```

lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make
Scanning dependencies of target websockets
[  1%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  2%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  3%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  5%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[  7%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 11%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 12%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 16%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 18%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
[ 20%] Building C object CMakeFiles/websockets.dir/lib/ssl-client.c.o
[ 21%] Building C object CMakeFiles/websockets.dir/lib/sha-1.c.o
[ 22%] Building C object CMakeFiles/websockets.dir/lib/lws-plat-unix.c.o
[ 24%] Building C object CMakeFiles/websockets.dir/lib/server.c.o
[ 25%] Building C object CMakeFiles/websockets.dir/lib/server-handshake.c.o
[ 26%] Building C object CMakeFiles/websockets.dir/lib/extension.c.o
[ 27%] Building C object CMakeFiles/websockets.dir/lib/extension-permessage-deflate.c.o
[ 29%] Building C object CMakeFiles/websockets.dir/lib/ranges.c.o
[ 30%] Linking C static library lib/libwebsockets.a
[ 30%] Built target websockets
Scanning dependencies of target test-client
[ 31%] Building C object CMakeFiles/test-client.dir/test-server/test-client.c.o
make[2]: *** No rule to make target '/opt/kamoebagateway/wolfssl/lib/libwolfssl.so', needed by 'bin/libwebsockets-test-client'.  Stop.
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/test-client.dir/all' failed
make[1]: *** [CMakeFiles/test-client.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # 
```

**Comment by lws-team** [comment]
`make V=1 VERBOSE=1` might let us see what it's actually trying to do.

**Comment by lyntongrice** [comment]
This help at all?
```

lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make v=1 VERBOSE=1
/usr/bin/cmake -H/home/Lynton/Desktop/KamoebaGateway/libwebsockets -B/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/websockets.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: Nothing to be done for 'CMakeFiles/websockets.dir/build'.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
[ 30%] Built target websockets
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/test-client.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: *** No rule to make target '/opt/kamoebagateway/wolfssl/lib/libwolfssl.so', needed by 'bin/libwebsockets-test-client'.  Stop.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/test-client.dir/all' failed
make[1]: *** [CMakeFiles/test-client.dir/all] Error 2
make[1]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # 
```

**Comment by lws-team** [comment]
No... needs to be capital 'V=1'

**Comment by lyntongrice** [comment]
Like this? 
```

lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make V=1 VERBOSE=1
/usr/bin/cmake -H/home/Lynton/Desktop/KamoebaGateway/libwebsockets -B/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/websockets.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: Nothing to be done for 'CMakeFiles/websockets.dir/build'.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
[ 30%] Built target websockets
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/test-client.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: *** No rule to make target '/opt/kamoebagateway/wolfssl/lib/libwolfssl.so', needed by 'bin/libwebsockets-test-client'.  Stop.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/test-client.dir/all' failed
make[1]: *** [CMakeFiles/test-client.dir/all] Error 2
make[1]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # 
```

**Comment by lws-team** [comment]
Can you do a `make clean`, or delete and regenerate the build directory and retry.

**Comment by lyntongrice** [comment]
Hi there,

BTW: Sorry for the trouble on this....I really appreciate your help.

I removed the entire build directory and re-created it doing this:

```
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/opt/kamoebagateway/libwebsockets/ -DLIB_SUFFIX=64 -DLWS_USE_CYASSL=0 -DLWS_USE_WOLFSSL=1 -DLWS_WOLFSSL_INCLUDE_DIRS=/opt/kamoebagateway/wolfssl/include -DLWS_WOLFSSL_LIBRARIES=/opt/kamoebagateway/wolfssl/lib/libwolfssl.so
```

But for some reason still get this:

```
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # make V=1 VERBOSE=1
/usr/bin/cmake -H/home/Lynton/Desktop/KamoebaGateway/libwebsockets -B/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/websockets.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/websockets.dir/build.make CMakeFiles/websockets.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: Nothing to be done for 'CMakeFiles/websockets.dir/build'.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
[ 30%] Built target websockets
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/depend
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
cd /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build /home/Lynton/Desktop/KamoebaGateway/libwebsockets/build/CMakeFiles/test-client.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make -f CMakeFiles/test-client.dir/build.make CMakeFiles/test-client.dir/build
make[2]: Entering directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
make[2]: *** No rule to make target '/opt/kamoebagateway/wolfssl/lib/libwolfssl.so', needed by 'bin/libwebsockets-test-client'.  Stop.
make[2]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/test-client.dir/all' failed
make[1]: *** [CMakeFiles/test-client.dir/all] Error 2
make[1]: Leaving directory '/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build'
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
lynton:/home/Lynton/Desktop/KamoebaGateway/libwebsockets/build # 
```

**Comment by lws-team** [comment]
This got solved on the cocos2d bug mentioned above by the user including the library in the link.

**Comment by lyntongrice** [comment]
Hi there,

Many thanks for this -- I just downloaded the latest "master" and it seemed to compile fine with wolfssl now ;-)

```
lynton:/home/Lynton/Desktop/test libwebsockets/libwebsockets-master/build # cmake .. -DCMAKE_INSTALL_PREFIX:PATH=/opt/kamoebagateway/libwebsockets/ -DLIB_SUFFIX=64 -DLWS_USE_CYASSL=0 -DLWS_USE_WOLFSSL=1 -DLWS_WOLFSSL_INCLUDE_DIRS=/opt/kamoebagateway/wolfssl/include -DLWS_WOLFSSL_LIBRARIES=/opt/kamoebagateway/wolfssl/lib/libwolfssl.so
-- The C compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_TOOLCHAIN_FILE=''
-- Found Git: /usr/bin/git  
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
Git commit hash: root@lynton-
-- Performing Test LWS_HAVE_inline
-- Performing Test LWS_HAVE_inline - Success
-- Performing Test LWS_HAVE___inline__
-- Performing Test LWS_HAVE___inline__ - Success
-- Performing Test LWS_HAVE___inline
-- Performing Test LWS_HAVE___inline - Success
-- Looking for bzero
-- Looking for bzero - found
-- Looking for fork
-- Looking for fork - found
-- Looking for getenv
-- Looking for getenv - found
-- Looking for malloc
-- Looking for malloc - found
-- Looking for memset
-- Looking for memset - found
-- Looking for realloc
-- Looking for realloc - found
-- Looking for socket
-- Looking for socket - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for vfork
-- Looking for vfork - found
-- Looking for execvpe
-- Looking for execvpe - found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for snprintf
-- Looking for snprintf - found
-- Looking for _snprintf
-- Looking for _snprintf - not found
-- Looking for _vsnprintf
-- Looking for _vsnprintf - not found
-- Looking for getloadavg
-- Looking for getloadavg - found
-- Looking for dlfcn.h
-- Looking for dlfcn.h - found
-- Looking for fcntl.h
-- Looking for fcntl.h - found
-- Looking for in6addr.h
-- Looking for in6addr.h - not found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for netinet/in.h
-- Looking for netinet/in.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stdlib.h
-- Looking for stdlib.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for string.h
-- Looking for string.h - found
-- Looking for sys/prctl.h
-- Looking for sys/prctl.h - found
-- Looking for sys/socket.h
-- Looking for sys/socket.h - found
-- Looking for sys/sockio.h
-- Looking for sys/sockio.h - not found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for vfork.h
-- Looking for vfork.h - not found
-- Looking for zlib.h
-- Looking for zlib.h - found
-- Looking for 4 include files stdlib.h, ..., float.h
-- Looking for 4 include files stdlib.h, ..., float.h - found
-- Performing Test LWS_HAVE_VISIBILITY
-- Performing Test LWS_HAVE_VISIBILITY - Success
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.8") 
zlib include dirs: /usr/include
zlib libraries: /usr/lib64/libz.so
Compiling with SSL support
wolfSSL include dir: /opt/kamoebagateway/wolfssl/include
wolfSSL libraries: /opt/kamoebagateway/wolfssl/lib/libwolfssl.so
-- Looking for SSL_CTX_set1_param
-- Looking for SSL_CTX_set1_param - not found
-- Looking for X509_VERIFY_PARAM_set1_host
-- Looking for X509_VERIFY_PARAM_set1_host - not found
 GENCERTS = 0
-- Looking for RPMTools... - rpmbuild NOT FOUND
---------------------------------------------------------------------
  Settings:  (For more help do cmake -LH <srcpath>)
---------------------------------------------------------------------
 LWS_WITH_STATIC = ON
 LWS_WITH_SHARED = ON
 LWS_WITH_SSL = ON (SSL Support)
 LWS_SSL_CLIENT_USE_OS_CA_CERTS = 1
 LWS_USE_WOLFSSL = 1 (wolfSSL/CyaSSL replacement for OpenSSL)
   LWS_WOLFSSL_LIBRARIES = /opt/kamoebagateway/wolfssl/lib/libwolfssl.so
   LWS_WOLFSSL_INCLUDE_DIRS = /opt/kamoebagateway/wolfssl/include
 LWS_WITHOUT_BUILTIN_SHA1 = OFF
 LWS_WITHOUT_BUILTIN_GETIFADDRS = OFF
 LWS_WITHOUT_CLIENT = OFF
 LWS_WITHOUT_SERVER = OFF
 LWS_LINK_TESTAPPS_DYNAMIC = OFF
 LWS_WITHOUT_TESTAPPS = OFF
 LWS_WITHOUT_TEST_SERVER = OFF
 LWS_WITHOUT_TEST_SERVER_EXTPOLL = OFF
 LWS_WITHOUT_TEST_PING = OFF
 LWS_WITHOUT_TEST_ECHO = OFF
 LWS_WITHOUT_TEST_CLIENT = OFF
 LWS_WITHOUT_TEST_FRAGGLE = OFF
 LWS_WITHOUT_EXTENSIONS = OFF
 LWS_WITH_LATENCY = OFF
 LWS_WITHOUT_DAEMONIZE = ON
 LWS_USE_LIBEV = 
 LWS_USE_LIBUV = 
 LWS_IPV6 = OFF
 LWS_UNIX_SOCK = OFF
 LWS_WITH_HTTP2 = OFF
 LWS_SSL_SERVER_WITH_ECDH_CERT = OFF
 LWS_MAX_SMP = 32
 LWS_WITH_CGI = OFF
 LWS_HAVE_OPENSSL_ECDH_H = 
 LWS_HAVE_SSL_CTX_set1_param = 
 LWS_WITH_HTTP_PROXY = OFF
 LIBHUBBUB_LIBRARIES = 
 PLUGINS = 
 LWS_WITH_ACCESS_LOG = OFF
 LWS_WITH_SERVER_STATUS = OFF
 LWS_WITH_LEJP = OFF
 LWS_WITH_LEJP_CONF = OFF
 LWS_WITH_SMTP = OFF
 LWS_WITH_GENERIC_SESSIONS = OFF
 LWS_STATIC_PIC = OFF
 LWS_WITH_RANGES = ON
 LWS_PLAT_OPTEE = OFF
 LWS_WITH_ESP32 = OFF
 LWS_WITH_ZIP_FOPS = ON
---------------------------------------------------------------------
-- Configuring done
-- Generating done
-- Build files have been written to: /home/Lynton/Desktop/test libwebsockets/libwebsockets-master/build
```
